### ISSUE TITLE
[FEATURE] make_outlier_scorer 

### ISSUE THREAD

**Issue by koaning** [koaning/scikit-lego#323]
From [this](https://koaning.io/posts/outliers-selection-vs-detection/) blogpost. It would be nice if we have a variant for "make_scorer" that can take metrics meant for classification and allow them to also be used to measure outlier detection algorithms.

```python
from sklearn.metrics import precision_score, recall_score
from sklearn.model_selection import GridSearchCV

df_subset = df.sort_values('class', ascending=False)[:80000]
X_subset = df_subset.drop(columns=['amount', 'class', 'time']).values
y_subset = df_subset['class']

def outlier_precision(mod, X, y):
    preds = mod.predict(X)
    return precision_score(y, np.where(preds == 1, 0, 1))

def outlier_recall(mod, X, y):
    preds = mod.predict(X)
    return recall_score(y, np.where(preds == 1, 0, 1))

forest = IsolationForest(contamination=0.1, behaviour="new", max_features=0.2)

mod = GridSearchCV(estimator=forest, 
                   cv=5,
                   n_jobs=-1,
                   scoring={"precision": outlier_precision,
                            "recall": outlier_recall}, 
                   refit="precision",
                   param_grid={'contamination': np.linspace(0.0001, 0.02, 30)})

mod.fit(X, y)
```

**Comment by koaning** [comment]
This might be better as a Meta model now that I think about it. 

@MBrouns, can I go ahead or do you have concerns?

**Comment by koaning** [comment]
This is my idea now. Instead of this; 

```python
from sklearn.metrics import precision_score, recall_score
from sklearn.model_selection import GridSearchCV

df_subset = df.sort_values('class', ascending=False)[:80000]
X_subset = df_subset.drop(columns=['amount', 'class', 'time']).values
y_subset = df_subset['class']

def outlier_precision(mod, X, y):
    preds = mod.predict(X)
    return precision_score(y, np.where(preds == 1, 0, 1))

def outlier_recall(mod, X, y):
    preds = mod.predict(X)
    return recall_score(y, np.where(preds == 1, 0, 1))

forest = IsolationForest(contamination=0.1, behaviour="new", max_features=0.2)

mod = GridSearchCV(estimator=forest, 
                   cv=5,
                   n_jobs=-1,
                   scoring={"precision": outlier_precision,
                            "recall": outlier_recall}, 
                   refit="precision",
                   param_grid={'contamination': np.linspace(0.0001, 0.02, 30)})

mod.fit(X, y)
```

You might be able to do; 

```python
from sklearn.metrics import precision_score, recall_score, make_score
from sklearn.model_selection import GridSearchCV
from sklego.meta import OutlierToClassifier

df_subset = df.sort_values('class', ascending=False)[:80000]
X_subset = df_subset.drop(columns=['amount', 'class', 'time']).values
y_subset = df_subset['class']

forest = IsolationForest(contamination=0.1, behaviour="new", max_features=0.2)

mod = GridSearchCV(estimator=OutlierToClassifier(forest), 
                   cv=5,
                   n_jobs=-1,
                   scoring={"precision": make_score(outlier_precision),
                          "recall": make_score(outlier_recall)}, 
                   refit="precision",
                   param_grid={'contamination': np.linspace(0.0001, 0.02, 30)})

mod.fit(X, y)
```

**Comment by MBrouns** [comment]
Hm that's an interesting question, which is slightly difficult to answer given that I don't really understand the rationale behind making outlier detectors return -1 and 1 in sklearn. Both approaches have their merit I think:

The meta-model approach makes it behave like a classifier, which is probably what you want in most cases anyway. 

Let's say though that at some point sklearn might introduce outlier detection specific scoring functions which rely on the -1 and 1 labels. In that case the scoring wrapper would provide more flexibility.

**Comment by koaning** [comment]
I think the choice for `-1` is because of algorithms like DBSCAN that do both clustering and outlier detection in one go. We could also both make the meta model as well as a make_outlier_scorer. Give the user both options?

**Comment by koaning** [comment]
I've implemented the meta model now. It feels like that is the easier method for now.
